packages/mds-utils/utils.ts

```diff
@@ -502,37 +502,73 @@ function areThereCommonElements<T, U>(arr1: T[], arr2: U[]) {
+ *      that the vehicle entered the reserved state after reservation_start,
+ *      even if the final state of the event is on_trip, and the rule will match.
+ *
+ * @example Matching transient `event_type`
+ * ```typescript
+ * isInStatesOrEvents({ states: { reserved: [] } }, { event_types: ['trip_end', 'reservation_start', 'trip_start'], vehicle_state: 'on_trip' }) => true
+ * ```
+ * @example State matching for transient `event_type` 'off_hours', but `event_type` not matched with explicit `event_type` in rule
+ * ```typescript
+ * isInStatesOrEvents({ states: { non_operational: ['maintenance'] } }, { event_types: ['trip_end', 'off_hours', 'on_hours'], vehicle_state: 'available' }) => false
+ * ```
+ * @example Match for last `event_type` and encoded `vehicle_state`, with explicit `event_type` in rule
+ * ```typescript
+ * isInStatesOrEvents({ states: { available: ['on_hours'] } }, { event_types: ['trip_end', 'off_hours', 'on_hours'], vehicle_state: 'available' }) => true
+ * ```
+ * @example Match for last `event_type` and encoded `vehicle_state`, with catch-all in rule
+ * ```typescript
+ * isInStatesOrEvents({ states: { available: [] } }, { event_types: ['on_hours'], vehicle_state: 'available' }) => true
+ * ```
+ */
+function isInStatesOrEvents(
+  rule: Pick<Rule, 'states'>,
+  event: Pick<VehicleEvent, 'event_types' | 'vehicle_state'>
+): boolean {
   const { states } = rule
   // If no states are specified, then the rule applies to all VehicleStates.
   if (states === null || states === undefined) {
     return true
   }
 
-  // States that it is possible to transition into with event.event_type
-  const possibleStates: VEHICLE_STATE[] = event.event_types.reduce((acc: VEHICLE_STATE[], event_type) => {
-    return acc.concat(EVENT_STATES_MAP[event_type])
-  }, [])
-  // The last element, assuming the provider didn't make a mistake, should be equivalent
-  // to the state of the event.
-  possibleStates.pop()
-  const result = possibleStates.reduce((acc, state) => {
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const matchableEvents: any = states[state as VEHICLE_STATE]
-    /* If there's a match between the event_type's transitionable events, and the
-     rule doesn't specify any events, or if there is a match between the rule and the specified
-     events, the rule matches this event. e.g. if the rule says { `available`: [`comms_lost`]} or
-     { `available`: [] }, it would match an event with event_type `comm_lost`.
-    */
+  const { event_types, vehicle_state } = event
+
+  // All event_types except the last (in most cases this will be an empty list)
+  const transientEventTypes = event_types.splice(0, -1)
+
+  /**
+   * State encoded in the event payload (default acc in the reducer) + states that it is
+   * possible to transition into with any transient event_types
+   */
+  const possibleStates: VEHICLE_STATE[] = transientEventTypes.reduce(
```

Comment:
So if you have `['comms_lost', 'comms_restored', 'comms_lost', 'comms_restored']`, you could have any possible state in between the first `comms_restored` and the `comms_lost` element, and the policy rule could match. Is this a sufficiently unlikely case that it's not worth worrying about?

Reply:
Tbh, I think that the MDS spec needs to be updated such that event_types which can map to multiple states _cannot_ be transient event_types. But yes, those ambiguous states in general present edge cases that could result in pretty weird Compliance results.

Reply:
Yeah, I'd be in favor of adding that to the spec, but it does make the whole thing more complicated.